packages/protocol/contracts/layer1/core/impl/Inbox.sol

```diff
@@ -372,228 +299,273 @@ contract Inbox is IInbox, IForcedInclusionStore, EssentialContract {
-        );
+        uint256 proposalsProven = _input.proposals.length - result_.firstProvenIndex;
+        if (proposalsProven == 1) {
+            result_.proposalAge = block.timestamp - uint256(firstReadyTimestamp);
+        }
 
-        ProvedEventPayload memory payload = ProvedEventPayload({
-            proposalId: _proposalId,
-            transition: _transition,
-            transitionRecord: _transitionRecord,
-            metadata: _metadata
-        });
-        emit Proved(_encodeProvedEventData(payload));
+        result_.aggregatedProvingHash = _hashTransitions(_input.transitions);
     }
 
-    /// @dev Persists transition record metadata in storage.
-    /// Returns false when an identical record already exists, avoiding redundant event
-    /// emissions.
-    /// @param _proposalId The proposal identifier.
-    /// @param _parentTransitionHash Hash of the parent transition for uniqueness.
-    /// @param _recordHash The keccak hash representing the transition record.
-    /// @param _hashAndDeadline The finalization metadata to store alongside the hash.
-    function _storeTransitionRecord(
-        uint48 _proposalId,
-        bytes32 _parentTransitionHash,
-        bytes26 _recordHash,
-        TransitionRecordHashAndDeadline memory _hashAndDeadline
+    /// @dev Processes sequential proofs, updates state, and builds the transition record.
+    /// @param _stateBefore The state before the proof is processed
+    /// @param _input The input containing the proposals and transitions
+    /// @return newState_ The new state after the proof is processed
+    /// @return bondInstruction_ Bond instruction to be signaled to L2 (BondType.NONE when unused)
+    /// @return firstReadyTimestamp_ The timestamp of the first ready proposal
+    function _processProof(
+        CoreState memory _stateBefore,
+        ProveInput memory _input
     )
-        internal
-        virtual
+        private
+        returns (
+            CoreState memory newState_,
+            LibBonds.BondInstruction memory bondInstruction_,
+            uint48 firstReadyTimestamp_,
+            uint256 firstProvenIndex_
+        )
     {
-        bytes32 compositeKey = _composeTransitionKey(_proposalId, _parentTransitionHash);
-        TransitionRecordHashAndDeadline storage entry =
-            _transitionRecordHashAndDeadline[compositeKey];
-        bytes26 recordHash = entry.recordHash;
-
-        if (recordHash == 0) {
-            entry.recordHash = _recordHash;
-            entry.finalizationDeadline = _hashAndDeadline.finalizationDeadline;
-        } else if (recordHash == _recordHash) {
-            emit TransitionDuplicateDetected();
-        } else {
-            emit TransitionConflictDetected();
-            conflictingTransitionDetected = true;
-            entry.finalizationDeadline = type(uint48).max;
+        unchecked {
+            newState_ = _stateBefore;
```

Comment:
maybe you don't have to make a copy. Updating _stateBefore directly will be fine.